### ISSUE TITLE
Kubectl API discovery caching may be up to 10 minutes stale

### ISSUE THREAD

**Issue by mbohlool** [kubernetes/kubernetes#47977]
kubectl caches API discovery results without being aware of its changes. Registering/Removing a new API service with api aggregator results in a change in API discovery but kubectl has no way to know the result is changed.

Possible solutions:

short term - implement e-tag for API discovery
long term - move to OpenAPI for api discovery

This is possibly a release blocker.

cc @lavalamp @pwittrock @dchen1107

@kubernetes/sig-cli-misc @kubernetes/sig-api-machinery-misc

**Comment by liggitt** [comment]
it's not ideal, but this is no different than how clients behaved with dynamic discovery changes due to TPR in any prior release, so I don't consider it a release blocker.

caches have a lifetime of 5 minutes, and can be cleared client-side.

**Comment by deads2k** [comment]
What are you doing to trigger this problems?  `kubectl get foo` check its cache, fails, re-primes the cache, checks again, succeeds, all silently to the user.

**Comment by madhusudancs** [comment]
What should other clients that are not kubectl do? This is a breaking change for kubefed for example. I am going to add the release-blocker label back, until we decide it is not for clients that are not kubectl.

**Comment by madhusudancs** [comment]
Btw, I forgot to mention that kubefed uses kubectl's cmd libraries.

**Comment by liggitt** [comment]
again, this behavior is no different than prior releases with dynamically added/removed TPR APIs

**Comment by dchen1107** [comment]
@liggitt Do we understand why this cause ci-kubernetes-e2e-gce-federation failing https://github.com/kubernetes/kubernetes/issues/47737?

**Comment by liggitt** [comment]
To my knowledge, nothing has changed in this area anytime recently. Didn't the Federation job just start failing a few days ago?

**Comment by caesarxuchao** [comment]
If the autoregister-completion health check is not done, the /healthz will return 500. can we let kubectl check the /healthz before visiting the discovery endpoints?

```
I0623 02:53:30.095497       5 wrap.go:42] GET /healthz: (571.071µs) 500
...[-]autoregister-completion failed: reason withheld\n
```

**Comment by deads2k** [comment]
This seems like much ado about a problem we've had since 1.2 (I think) when we got TPRs. The discovery doc changes in response to those and clients already had to deal with it.  The same thing happens on upgrades and even server restarts since we added API groups years ago. 

Claiming this as a new problem is a mischaracterisation since we've had it for years.  In addition, the exact case described here is incorrect (kubectl live checks on a miss).

**Comment by dchen1107** [comment]
@deads2k I am not sure about when we introduced the issue here. But what you described for upgrade is matching what I observed during upgrade test debugging for 1.7. 

The only reason we mark this as the release blocker is the federation team think this is the root cause for https://github.com/kubernetes/kubernetes/issues/47737. We need to find the solution / workaround / document as the known issue for federation / upgrade / etc.. The long term fixing is not required here.

**Comment by deads2k** [comment]
> @deads2k I am not sure about when we introduced the issue here. But what you described for upgrade is matching what I observed during upgrade test debugging for 1.7.

@dchen1107 Then I would avoid trying to make rushed API changes to discovery or behavior modifications to kubectl.  The aggregation behavior people seem focused on merged in March with this pull here https://github.com/kubernetes/kubernetes/pull/42911 .

**Comment by lavalamp** [comment]
This is a behavior change on the server that exposes a caching bug (which itself is probably both a server and client bug).

There is no reason for apiserver to return a partial list of built-in resources at any point in its startup stack. I consider that a bug, it is exposing internal setup details to the world for no good reason. Since it is breaking clients and we aren't sure the etag is set correctly to allow caches to work, I think we need to fix the server bug.

Caching for X amount of time is also clearly really wrong now that we are super dynamic, the server must present an ETag and the client needs to use the If-None-Match header to do it right.

**Comment by mbohlool** [comment]
In the Burndown meeting, the decision was not to fix this, only document it as a known issue.

**Comment by liggitt** [comment]
the kubectl and kubefed cache usage will be fixed

**Comment by dchen1107** [comment]
xref: v1.7 known issue: https://github.com/kubernetes/kubernetes/issues/46733

**Comment by marun** [comment]
@liggitt Given that the pr's you've linked to have all merged, should this issue continue blocking 1.7?  If not, please remove the 'approved-for-milestone' label.

**Comment by liggitt** [comment]
The api-versions and kubefed cache usage has been fixed for 1.7

**Comment by liggitt** [comment]
issue tracking caching headers / etag support is https://github.com/kubernetes/kubernetes/issues/44957
